### PR TITLE
Add DebugMap method to generated types

### PR DESCRIPTION
**Pull request by josephschorr**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -261,6 +265,74 @@ func writeNewXWithOptionsAndDefaults(buf *jen.File, c Config) {
 	})
 }
 
+var (
+	DebugMapFieldTag = "debugmap"
+	SensitiveNames   = []string{"secure"}
+)
+
+func writeDebugMap(buf *jen.File, st *types.Struct, c Config) {
+	newFuncName := fmt.Sprintf("DebugMap")
```
Comment (ecordell) [inline main.go:285]: What do you think about doing this as [GoString](https://pkg.go.dev/fmt#GoStringer)? 

That should make it a lot harder to accidentally print sensitive fields

File: main.go [hunk 2]
```diff
@@ -261,6 +265,74 @@ func writeNewXWithOptionsAndDefaults(buf *jen.File, c Config) {
 	})
 }
 
+var (
+	DebugMapFieldTag = "debugmap"
```
Comment (ecordell) [inline main.go:281]: nit: const

File: main.go [hunk 3]
```diff
@@ -261,6 +265,74 @@ func writeNewXWithOptionsAndDefaults(buf *jen.File, c Config) {
...
+				panic(err)
+			}
+
+			tag, err := tags.Get(DebugMapFieldTag)
+			if err != nil {
+				fmt.Printf("missing debugmap tag on field %s in type %s\n", f.Name(), c.TargetTypeName)
+				os.Exit(1)
+			}
+
+			switch tag.Value() {
```
Comment (ecordell) [inline main.go:309]: Do we need `structtag`? Can't you just do:

```suggestion
			switch f.Tag.Get(DebugMapFieldTag) {
```

File: main.go [hunk 4]
```diff
@@ -261,6 +265,74 @@ func writeNewXWithOptionsAndDefaults(buf *jen.File, c Config) {
...
+						fmt.Printf("field %s in type %s must be marked as 'sensitive'\n", f.Name(), c.TargetTypeName)
+						os.Exit(1)
+					}
+				}
+
+				grp.Id(mapId).Index(jen.Lit(f.Name())).Op("=").Qual("github.com/ecordell/optgen/helpers", "DebugValue").Call(jen.Id(c.ReceiverId).Dot(f.Name()), jen.Lit(false))
+
+			case "visible-format":
+				for _, sensitiveName := range SensitiveNames {
+					if strings.Contains(strings.ToLower(f.Name()), sensitiveName) {
```
Comment (ecordell) [inline main.go:322]: this should probably be configurable via the cli args - not everyone will necessarily agree on sensitivenames

File: main.go (hunk 1 above)
Comment (josephschorr) [reply main.go:285]: It needs to be a map to make it easy for us to log via zerolog

File: main.go (hunk 2 above)
Comment (josephschorr) [reply main.go:281]: I had it `const`, but the next var can't be, so I just combined them

File: main.go (hunk 3 above)
Comment (josephschorr) [reply main.go:309]: Screen Shot 2023-05-08 at 4 30 39 PM

Comment (josephschorr) [reply main.go:309]: Screen Shot 2023-05-08 at 4 31 00 PM

File: main.go (hunk 4 above)
Comment (josephschorr) [reply main.go:322]: I think its still worth having a set of defaults

File: main.go (hunk 3 above)
Comment (josephschorr) [reply main.go:309]: Also doesn't work on `st.Tag`, which is just a string

**Comment by josephschorr** [comment]
Updated

**Review by ecordell — APPROVED** [review]

**Pull request merged**